druid/src/widget/wgpu_view.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright 2018 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A view hosting a wgpu instance for 3D rendering.
+//!
+//! This widget has several limitations compared to other standard widget due
+//! to the way it interfaces with wgpu via its own native window, instead of
+//! reusing the same top-level native window like other widgets do. In particular
+//! this means it can only paint within its bounds and cannot be painted over by
+//! any other widget, and therefore does not support being embedded into other
+//! widgets like scroll views which partially obscure a widget's client area.
+
+#![cfg(feature = "wgpu_view")]
+
+use crate::widget::prelude::*;
+use crate::{Data, NativeWindowHandle};
+
+use log::{debug, info};
+
+/// A trait for rendering custom content into a `WgpuView` by accessing its internal wgpu pipeline.
+pub trait WgpuRenderer {
+    /// Callback invoked to initilize the painter once the wgpu device and swap chain are up.
+    fn init(
+        &mut self,
+        sc_desc: &wgpu::SwapChainDescriptor,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+    );
+    /// Callback invoked when the wgpu view is resized.
+    ///
+    /// This is typically used to update any camera projection matrix, and/or recreate any render
+    /// target. This is invoked before the new swap chain with the new window size is created, but
+    /// after the size has been updated in `sc_desc`.
+    fn resize(
+        &mut self,
+        sc_desc: &wgpu::SwapChainDescriptor,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+    );
+    /// Callback invoked when the wgpu view needs to be repaint.
+    ///
+    /// This is where rendering must be submitted by the painter to the given queue.
+    fn render(
+        &mut self,
+        frame: &wgpu::SwapChainTexture,
+        device: &wgpu::Device,
+        queue: &wgpu::Queue,
+    );
+}
```

Comment:
This trait seems like it should have event, lifecycle and update methods (that pass along the references to data and env) to allow interactivity and reactivity.

Reply:
You're talking about the native window events (mouse click _etc._) for example? That makes sense I guess.

Reply:
I mean druid events, so someone can effectively write a widget here all in one place. I think in the end probably WgpuRenderer is a WgpuWidgetHost, ie it adapts a WgpuWidget into a druid one. I can put a proposal together as another PR on your repo, might be easier to explain what I mean in code...